File: vane-core/src/main/java/org/oddlama/vane/core/config/ConfigItemStackField.java
package org.oddlama.vane.core.config;

import static org.oddlama.vane.util.MaterialUtil.materialFrom;
import static org.oddlama.vane.util.StorageUtil.namespacedKey;

import java.lang.reflect.Field;
import java.util.function.Function;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.inventory.ItemStack;
import org.oddlama.vane.annotation.config.ConfigItemStack;
import org.oddlama.vane.core.YamlLoadException;

public class ConfigItemStackField extends ConfigField<ItemStack> {

    public ConfigItemStack annotation;

    public ConfigItemStackField(
        Object owner,
        Field field,
        Function<String, String> mapName,
        ConfigItemStack annotation
    ) {
        super(owner, field, mapName, "item stack", annotation.desc());
        this.annotation = annotation;
    }

    private void appendItemStackDefinition(StringBuilder builder, String indent, String prefix, ItemStack def) {
        // Material
        builder.append(indent);
        builder.append(prefix);
        builder.append("  material: ");
        final var material =
            "\"" +
            escapeYaml(def.getType().getKey().getNamespace()) +
            ":" +
            escapeYaml(def.getType().getKey().getKey()) +
            "\"";
        builder.append(material);
        builder.append("\n");

        // Amount
        if (def.getAmount() != 1) {
            builder.append(indent);
            builder.append(prefix);
            builder.append("  amount: ");
            builder.append(def.getAmount());
            builder.append("\n");
        }
    }

    @Override
    public ItemStack def() {
        final var override = overriddenDef();
        if (override != null) {
            return override;
        } else {
            return new ItemStack(annotation.def().type(), annotation.def().amount());
        }
    }

    @Override
    public boolean metrics() {
        final var override = overriddenMetrics();
        if (override != null) {
            return override;
        } else {
            return annotation.metrics();
        }
    }

    @Override
    public void generateYaml(StringBuilder builder, String indent, YamlConfiguration existingCompatibleConfig) {
        appendDescription(builder, indent);

        // Default
        builder.append(indent);
        builder.append("# Default:\n");
        appendItemStackDefinition(builder, indent, "# ", def());

        // Definition
        builder.append(indent);
        builder.append(basename());
        builder.append(":\n");
        final var def = existingCompatibleConfig != null && existingCompatibleConfig.contains(yamlPath())
            ? loadFromYaml(existingCompatibleConfig)
            : def();
        appendItemStackDefinition(builder, indent, "", def);
    }

    @Override
    public void checkLoadable(YamlConfiguration yaml) throws YamlLoadException {
        checkYamlPath(yaml);

        if (!yaml.isConfigurationSection(yamlPath())) {
            throw new YamlLoadException("Invalid type for yaml path '" + yamlPath() + "', expected group");
        }

        for (var varKey : yaml.getConfigurationSection(yamlPath()).getKeys(false)) {
            final var varPath = yamlPath() + "." + varKey;
            switch (varKey) {
                case "material": {
                    if (!yaml.isString(varPath)) {
                        throw new YamlLoadException("Invalid type for yaml path '" + varPath + "', expected list");
                    }

                    final var str = yaml.getString(varPath);
                    final var split = str.split(":");
                    if (split.length != 2) {
                        throw new YamlLoadException(
                            "Invalid material for yaml path '" +
                            yamlPath() +
                            "': '" +
                            str +
                            "' is not a valid namespaced key"
                        );
                    }
                    break;
                }
                case "amount": {
                    if (!(yaml.get(varPath) instanceof Number)) {
                        throw new YamlLoadException("Invalid type for yaml path '" + yamlPath() + "', expected int");
                    }
                    final var val = yaml.getInt(yamlPath());
                    if (val < 0) {
                        throw new YamlLoadException("Invalid value for yaml path '" + yamlPath() + "' Must be >= 0");
                    }
                    break;
                }
            }
        }
    }

    public ItemStack loadFromYaml(YamlConfiguration yaml) {
        var materialStr = "";
        var amount = 1;
        for (var varKey : yaml.getConfigurationSection(yamlPath()).getKeys(false)) {
            final var varPath = yamlPath() + "." + varKey;
            switch (varKey) {
                case "material": {
                    amount = 0;
                    materialStr = yaml.getString(varPath);
                    break;
                }
                case "amount": {
                    amount = yaml.getInt(varPath);
                    break;
                }
            }
        }

        final var split = materialStr.split(":");
        final var material = materialFrom(namespacedKey(split[0], split[1]));
        return new ItemStack(material, amount);
    }

    public void load(YamlConfiguration yaml) {
        try {
            field.set(owner, loadFromYaml(yaml));
        } catch (IllegalAccessException e) {
            throw new RuntimeException("Invalid field access on '" + field.getName() + "'. This is a bug.");
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/config/ConfigItemStackField.kt
package org.oddlama.vane.core.config

import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.inventory.ItemStack
import org.oddlama.vane.annotation.config.ConfigItemStack
import org.oddlama.vane.core.YamlLoadException
import org.oddlama.vane.util.MaterialUtil.materialFrom
import org.oddlama.vane.util.StorageUtil.namespacedKey
import java.lang.reflect.Field

/**
 * Config field handler for item stack values.
 *
 * @param owner object containing the reflected config field.
 * @param field reflected config field.
 * @param mapName maps Java field names to YAML paths.
 * @param annotation source annotation metadata.
 */
class ConfigItemStackField(
    owner: Any?,
    field: Field,
    mapName: (String?) -> String?,
    /** Annotation metadata for this field. */
    var annotation: ConfigItemStack
) : ConfigField<ItemStack?>(owner, field, mapName, "item stack", annotation.desc) {

    /** Formats an item stack material as a quoted namespaced key. */
    private fun ItemStack.materialKeyString() =
        "\"${escapeYaml(type.key.namespace)}:${escapeYaml(type.key.key)}\""

    /** Appends an item stack definition block. */
    private fun appendItemStackDefinition(builder: StringBuilder, indent: String?, prefix: String?, def: ItemStack) {
        builder.append("$indent$prefix  material: ${def.materialKeyString()}\n")
        if (def.amount != 1) builder.append("$indent$prefix  amount: ${def.amount}\n")
    }

    /** Returns the default value for this config field. */
    override fun def(): ItemStack = overriddenDef() ?: ItemStack(annotation.def.type, annotation.def.amount)

    /** Returns whether metrics collection is enabled for this field. */
    override fun metrics(): Boolean = overriddenMetrics() ?: annotation.metrics

    /** Generates YAML for this field. */
    override fun generateYaml(builder: StringBuilder, indent: String, existingCompatibleConfig: YamlConfiguration?) {
        appendDescription(builder, indent)
        builder.append("$indent# Default:\n")
        appendItemStackDefinition(builder, indent, "# ", def())
        builder.append("$indent${basename()}:\n")
        val def = if (existingCompatibleConfig != null && existingCompatibleConfig.contains(yamlPath()))
            loadFromYaml(existingCompatibleConfig) else def()
        appendItemStackDefinition(builder, indent, "", def)
    }

    /** Validates that this field is loadable from YAML. */
    @Throws(YamlLoadException::class)
    override fun checkLoadable(yaml: YamlConfiguration) {
        checkYamlPath(yaml)
        if (!yaml.isConfigurationSection(yamlPath()))
            throw YamlLoadException("Invalid type for yaml path '${yamlPath()}', expected group")
        for (varKey in yaml.getConfigurationSection(yamlPath())!!.getKeys(false)) {
            val varPath = "${yamlPath()}.$varKey"
            when (varKey) {
                "material" -> {
                    if (!yaml.isString(varPath))
                        throw YamlLoadException("Invalid type for yaml path '$varPath', expected list")
                    val str = yaml.getString(varPath)!!
                    val (ns, key) = requireNamespacedKeyParts(yamlPath(), str, "material")
                    if (ns.isEmpty() || key.isEmpty())
                        throw YamlLoadException("Invalid material for yaml path '${yamlPath()}': '$str' is not a valid namespaced key")
                }
                "amount" -> {
                    if (yaml.get(varPath) !is Number)
                        throw YamlLoadException("Invalid type for yaml path '${yamlPath()}', expected int")
                    if (yaml.getInt(varPath) < 0)
                        throw YamlLoadException("Invalid value for yaml path '${yamlPath()}' Must be >= 0")
                }
            }
        }
    }

    /** Loads this field value from YAML. */
    fun loadFromYaml(yaml: YamlConfiguration): ItemStack {
        var materialStr = ""
        var amount = 1
        for (varKey in yaml.getConfigurationSection(yamlPath())!!.getKeys(false)) {
            val varPath = "${yamlPath()}.$varKey"
            when (varKey) {
                "material" -> { materialStr = yaml.getString(varPath)!!; amount = 0 }
                "amount"   -> amount = yaml.getInt(varPath)
            }
        }
        val (ns, key) = requireNamespacedKeyParts(yamlPath(), materialStr, "material")
        return ItemStack(materialFrom(namespacedKey(ns, key))!!, amount)
    }

    /** Writes the loaded value into the reflected field. */
    override fun load(yaml: YamlConfiguration) {
        try {
            field.set(owner, loadFromYaml(yaml))
        } catch (_: IllegalAccessException) {
            throw RuntimeException("Invalid field access on '${field.name}'. This is a bug.")
        }
    }
}
